pgrx-examples/decoder/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+//
+
+#[pg_guard]
+unsafe extern "C" fn pg_decode_startup(
+    ctx_ptr: *mut pg_sys::LogicalDecodingContext,
+    options_ptr: *mut pg_sys::OutputPluginOptions,
+    _is_init: bool,
+) {
+    let mut options = unsafe { PgBox::from_pg(options_ptr) };
+    options.output_type = pg_sys::OutputPluginOutputType::OUTPUT_PLUGIN_TEXTUAL_OUTPUT;
+    options.into_pg();
+
+    // output_plugin_private is used to store a pointer to a common state shared
+    // between all the callbacks of the same transaction. There may be a better
+    // way to manage this pointer, but currently alloc/dealloc the cleanest way
+    // I could think of...
+    let layout = Layout::new::<DecodingState>();
+    let mut ctx = unsafe { PgBox::from_pg(ctx_ptr) };
+    ctx.output_plugin_private = unsafe { alloc(layout) as *mut std::ffi::c_void };
+    ctx.into_pg();
+}
+
+#[pg_guard]
+unsafe extern "C" fn pg_decode_begin_txn(
+    ctx_ptr: *mut pg_sys::LogicalDecodingContext,
+    _txn_ptr: *mut pg_sys::ReorderBufferTXN,
+) {
+    let ctx = unsafe { PgBox::from_pg(ctx_ptr) };
+    let mut state = unsafe { PgBox::from_pg(ctx.output_plugin_private as *mut DecodingState) };
+    state.xact_change_counter = 0;
+    Action::begin().output_plugin_write(ctx);
+}
+
+#[pg_guard]
+unsafe extern "C" fn pg_decode_commit_txn(
+    ctx_ptr: *mut pg_sys::LogicalDecodingContext,
+    txn_ptr: *mut pg_sys::ReorderBufferTXN,
+    _commit_lsn: pg_sys::XLogRecPtr,
+) {
+    let ctx = unsafe { PgBox::from_pg(ctx_ptr) };
+    let txn = unsafe { PgBox::from_pg(txn_ptr) };
+    let state = unsafe { PgBox::from_pg(ctx.output_plugin_private as *mut DecodingState) };
+    Action::commit(txn, state.xact_change_counter).output_plugin_write(ctx);
+}
+
+#[pg_guard]
+unsafe extern "C" fn pg_decode_change(
+    ctx_ptr: *mut pg_sys::LogicalDecodingContext,
+    _txn_ptr: *mut pg_sys::ReorderBufferTXN,
+    relation: pg_sys::Relation,
+    change_ptr: *mut pg_sys::ReorderBufferChange,
+) {
+    let ctx = unsafe { PgBox::from_pg(ctx_ptr) };
+    let mut state = unsafe { PgBox::from_pg(ctx.output_plugin_private as *mut DecodingState) };
+    state.xact_change_counter += 1;
+
+    let pgrelation = unsafe { pgrx::PgRelation::from_pg(relation) };
+    let change = unsafe { PgBox::from_pg(change_ptr) };
+
+    Action::change(pgrelation, change).output_plugin_write(ctx);
+}
+
+#[pg_guard]
+unsafe extern "C" fn pg_decode_shutdown(ctx_ptr: *mut pg_sys::LogicalDecodingContext) {
+    let layout = Layout::new::<DecodingState>();
+    let ctx = unsafe { PgBox::from_pg(ctx_ptr) };
+    unsafe { dealloc(ctx.output_plugin_private as *mut u8, layout) };
+}
+
+#[cfg(any(test, feature = "pg_test"))]
+#[pg_schema]
+mod tests {
+    use pgrx::prelude::*;
+
+    #[pg_test]
+    fn test_action_begin() {
+        use crate::Action;
+
+        let input = Action::begin();
+        let json = serde_json::to_string(&input).expect("Serde Error");
+        assert_eq!(json, "{\"typ\":\"BEGIN\"}");
```

Comment:
We don't necessarily need vigorous escaping if we use a raw string:
```suggestion
        assert_eq!(json, r#"{"typ":"BEGIN"}"#);
```